### PR TITLE
CSHARP-3918: Support authorizedCollections option for listCollections helpers.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
Waiting for EG

### CODE REVIEW

File: src/MongoDB.Driver.Core/Core/Operations/ListCollectionsOperation.cs [hunk 1]
```diff
@@ -26,9 +28,10 @@ namespace MongoDB.Driver.Core.Operations
     /// <summary>
     /// Represents a list collections operation.
     /// </summary>
-    public class ListCollectionsOperation : IReadOperation<IAsyncCursor<BsonDocument>>
+    public class ListCollectionsOperation : IReadOperation<IAsyncCursor<BsonDocument>>, IExecutableInRetryableReadContext<IAsyncCursor<BsonDocument>>
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Operations/ListCollectionsOperation.cs:31]: The diff in this file is mostly related to the fact that I merged ListCollectionsOperation and ListCollectionsUsingCommandOperation. Effectively I removed `ListCollectionsOperation` and renamed `ListCollectionsUsingCommandOperation` to `ListCollectionsOperation`

File: tests/MongoDB.Driver.Core.Tests/Core/Operations/ListCollectionsOperationTests.cs [hunk 2]
```diff
@@ -65,12 +61,24 @@ public void constructor_should_throw_when_messageEncoderSettings_is_null()
             action.ShouldThrow<ArgumentNullException>().And.ParamName.Should().Be("messageEncoderSettings");
         }
 
+        [Fact]
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Core.Tests/Core/Operations/ListCollectionsOperationTests.cs:64]: The diff in this file is mostly related to the fact that I merged ListCollectionsOperationTests and ListCollectionsUsingCommandOperationTests. Effectively I removed ListCollectionsOperationTests and renamed ListCollectionsUsingCommandOperationTests to ListCollectionsOperationTests

File: tests/MongoDB.Driver.Tests/MongoDatabaseImplTests.cs [hunk 3]
```diff
@@ -888,26 +892,32 @@ public void Settings_should_be_set()
             using (var session = usingSession ? client.StartSession() : null)
             {
                 IAsyncCursor<string> cursor;
+                var listCollectionNamesOptions = new ListCollectionNamesOptions();
+                if (usingAuthorizedCollection)
+                {
+                    listCollectionNamesOptions.AuthorizedCollections = true;
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Tests/MongoDatabaseImplTests.cs:898]: in fact server doesn't throw exception for any property here, for example the below works "fine" too:

    MongoDB Enterprise mongos> db.runCommand({ listCollections:1, fake:1})
    {
            "result" : [ ],
            "cursor" : {
                    "id" : NumberLong(0),
                    "ns" : "test.$cmd.listCollections",
                    "firstBatch" : [ ]
            },
            "ok" : 1,
            "operationTime" : Timestamp(1641840858, 1),
            "$clusterTime" : {
                    "clusterTime" : Timestamp(1641840858, 1),
                    "signature" : {
                            "hash" : BinData(0,"AAAAAAAAAAAAAAAAAAAAAAAAAAA="),
                            "keyId" : NumberLong(0)
                    }
            }
    }

so the fact that the server didn't return exception for `AuthorizedCollections:true` diesn't mean it works correctly, but I would probably not to worry about this and would not involve tests with auth workflow

File: tests/MongoDB.Driver.Tests/MongoDatabaseImplTests.cs [hunk 4]
```diff
@@ -970,10 +982,11 @@ public void Settings_should_be_set()
             var op = call.Operation.Should().BeOfType<ListCollectionsOperation>().Subject;
             op.DatabaseNamespace.Should().Be(_subject.DatabaseNamespace);
             op.NameOnly.Should().NotHaveValue();
-            if (usingFilter || usingBatchSize)
+            if (usingFilter || usingBatchSize || authorizedCollections.HasValue)
             {
                 op.Should().Match<ListCollectionsOperation>(
                     (o) =>
+                        o.AuthorizedCollections == (authorizedCollections.HasValue ? authorizedCollections.Value : null) &&
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Tests/MongoDatabaseImplTests.cs:989]: `o.AuthorizedCollections == authorizedCollections` should be sufficient

File: src/MongoDB.Driver/ListCollectionsOptions.cs [hunk 5]
```diff
@@ -23,10 +23,20 @@ namespace MongoDB.Driver
     public sealed class ListCollectionsOptions
     {
         // fields
+        private bool? _authorizedCollections;
```
Comment (BorisDog) [inline src/MongoDB.Driver/ListCollectionsOptions.cs:26]: Should `_authorizedCollections` option be added to `ListCollectionNamesOptions` only? As it has no effect without `nameOnly: true`, and we don't set `nameOnly` in this case. Or alternatively should we implement the "Drivers MUST specify true for the ``nameOnly`` option ..." section ?

**Review by BorisDog — CHANGES_REQUESTED** [review]

**Comment by rstam** [comment]
I added myself as a reviewer because I want to review any further changes that might happen with respect to the `nameOnly` flag (I'm trying to decipher the Slack discussion on that).

File: src/MongoDB.Driver/ListCollectionsOptions.cs (hunk 5 above)
Comment (rstam) [reply src/MongoDB.Driver/ListCollectionsOptions.cs:26]: I agree that it should be added for `ListCollectionNames` only.

**Review by rstam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Operations/ListCollectionsOperation.cs [hunk 6]
```diff
@@ -51,6 +54,18 @@ public class ListCollectionsOperation : IReadOperation<IAsyncCursor<BsonDocument
         }
 
         // properties
+        /// <summary>
+        /// Gets or sets the AuthorizedCollections flag.
+        /// </summary>
+        /// <value>
+        /// The authorizedCollections.
```
Comment (BorisDog) [inline src/MongoDB.Driver.Core/Core/Operations/ListCollectionsOperation.cs:61]: nit: Return value: `Whether authorizedCollections flag/option is set...`

File: tests/MongoDB.Driver.Tests/MongoDatabaseImplTests.cs [hunk 7]
```diff
@@ -866,13 +869,14 @@ public void Settings_should_be_set()
         [ParameterAttributeData]
         public void ListCollectionNames_should_return_expected_result(
             [Values(0, 1, 2, 10)] int numberOfCollections,
+            [Values(false, true)] bool usingAuthorizedCollection,
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Tests/MongoDatabaseImplTests.cs:872]: nit: usingAuthorizedCollection**s**

File: tests/MongoDB.Driver.Tests/MongoDatabaseImplTests.cs [hunk 8]
```diff
@@ -888,26 +892,32 @@ public void Settings_should_be_set()
             using (var session = usingSession ? client.StartSession() : null)
             {
                 IAsyncCursor<string> cursor;
+                var listCollectionNamesOptions = new ListCollectionNamesOptions();
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Tests/MongoDatabaseImplTests.cs:895]: The overload without options not tested anymore. Maybe we can add `null`  to `usingAuthorizedCollections `parameter values for that: `[Values(false, true, null)] bool usingAuthorizedCollection`?

**Review by BorisDog — CHANGES_REQUESTED** [review]
Minor comments

**Review by rstam — APPROVED** [review]
LGTM but please re-request my review if you make any further significant changes.

File: src/MongoDB.Driver.Core/Core/Operations/ListCollectionsOperation.cs (hunk 6 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Operations/ListCollectionsOperation.cs:61]: done

File: tests/MongoDB.Driver.Tests/MongoDatabaseImplTests.cs (hunk 7 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/MongoDatabaseImplTests.cs:872]: done

File: tests/MongoDB.Driver.Tests/MongoDatabaseImplTests.cs (hunk 8 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/MongoDatabaseImplTests.cs:895]: done

**Comment by DmitryLukyanov** [comment]
ready for review

**Review by BorisDog — APPROVED** [review]
LGTM

**Pull request closed without merging**